statement_parser.go

```diff
@@ -226,41 +235,67 @@ func findParams(sql string) ([]string, error) {
 			} else {
 				lastCharWasEscapeChar = false
 			}
+			parsedSQL.WriteRune(c)
 		} else {
 			// We are not in a quoted string. It's a parameter if it is an '@' followed by a letter or an underscore.
 			// See https://cloud.google.com/spanner/docs/lexical#identifiers for identifier rules.
 			if c == paramPrefix && len(runes) > index+1 && (unicode.IsLetter(runes[index+1]) || runes[index+1] == '_') {
+				if hasPositionalParameter {
+					return sql, nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "statement must not contain both named and positional parameter: %s", sql))
+				}
+				parsedSQL.WriteRune(c)
 				index++
 				startIndex := index
 				for index < len(runes) {
 					if !(unicode.IsLetter(runes[index]) || unicode.IsDigit(runes[index]) || runes[index] == '_') {
+						hasNamedParameter = true
 						res = append(res, string(runes[startIndex:index]))
+						hasNamedParameter = true
+						parsedSQL.WriteRune(runes[index])
 						break
 					}
 					if index == len(runes)-1 {
+						hasNamedParameter = true
 						res = append(res, string(runes[startIndex:]))
+						parsedSQL.WriteRune(runes[index])
 						break
 					}
+					parsedSQL.WriteRune(runes[index])
 					index++
 				}
+			} else if c == positionalParamChar {
+				if hasNamedParameter {
+					return sql, nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "statement must not contain both named and positional parameter: %s", sql))
+				}
+				hasPositionalParameter = true
+				parsedSQL.WriteString("@p" + strconv.Itoa(positionalParameterIndex))
+				res = append(res, "p"+strconv.Itoa(positionalParameterIndex))
+				positionalParameterIndex++
 			} else {
 				if c == singleQuote || c == doubleQuote || c == backtick {
 					isInQuoted = true
 					startQuote = c
 					// Check whether it is a triple-quote.
 					if len(runes) > index+2 && runes[index+1] == startQuote && runes[index+2] == startQuote {
 						isTripleQuoted = true
+						parsedSQL.WriteRune(c)
+						parsedSQL.WriteRune(c)
 						index += 2
 					}
 				}
+				parsedSQL.WriteRune(c)
 			}
 		}
 		index++
 	}
 	if isInQuoted {
-		return nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "statement contains an unclosed literal: %s", sql))
+		return sql, nil, spanner.ToSpannerError(status.Errorf(codes.InvalidArgument, "statement contains an unclosed literal: %s", sql))
+	}
+	sql = strings.TrimSpace(parsedSQL.String())
```

Comment:
Both for 'safety' and efficiency reasons, I think it would make sense here to only return the `parsedSQL` value if `hasPositionalParameters` is true. If `hasPositionalParameters` is false, we can just return the input `sql` value.